_posts/2019-06-04-flink-network-stack.md

```diff
@@ -0,0 +1,324 @@
+    <td class="tg-c3ow" colspan="2" rowspan="2">local</td>
+    <td class="tg-c3ow" colspan="2" rowspan="2">remote</td>
+  </tr>
+  <tr>
+    <td class="tg-0pky">A.2</td>
+  </tr>
+  <tr>
+    <td class="tg-0pky">A.3</td>
+    <td class="tg-c3ow" colspan="2" rowspan="2">remote</td>
+    <td class="tg-c3ow" colspan="2" rowspan="2">local</td>
+  </tr>
+  <tr>
+    <td class="tg-0pky">A.4</td>
+  </tr>
+</table>
+</center>
+
+<br>
+
+Each (remote) network connection between different tasks will get its own TCP channel in Flink’s network stack. However, if different subtasks of the same task are scheduled onto the same TaskManager, their network connections towards the same TaskManagers will be multiplexed and share a single TCP channel for reduced resource usage. In our example, this would apply to A.1 → B.3, A.1 → B.4, as well as A.2 → B.3, and A.2 → B.4 as pictured below:
+<br>
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack2.png" width="700px" alt="Physical-transport-Flink's Network Stack"/>
+</center>
+<br>
+
+The results of each subtask are called [ResultPartition](https://ci.apache.org/projects/flink/flink-docs-release-1.8/api/java/org/apache/flink/runtime/io/network/partition/ResultPartition.html), each split into separate [ResultSubpartitions](https://ci.apache.org/projects/flink/flink-docs-release-1.8/api/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.html) — one for each logical channel. At this point in the stack, Flink is not dealing with individual records anymore but instead with a group of serialised records assembled together into network buffers. The number of buffers available to each subtask in its own local buffer pool (one per sending and receiving side each) is limited to at most
+
+    #channels * buffers-per-channel + floating-buffers-per-gate
+
+The total number of buffers on a single TaskManager usually does not need configuration. See the [Configuring the Network Buffers](https://ci.apache.org/projects/flink/flink-docs-release-1.8/ops/config.html#configuring-the-network-buffers) documentation for details on how to do so if needed.
+
+Whenever a subtask’s sending buffer pool is exhausted — buffers reside in either a result subpartition's buffer queue or inside the lower, Netty-backed network stack — the producer is blocked, cannot continue, and experiences backpressure. The receiver works in a similar fashion: any incoming Netty buffer in the lower network stack needs to be made available to Flink via a network buffer. If there is no network buffer available in the appropriate subtask's buffer pool, Flink will stop reading from this channel until a buffer becomes available. This would effectively backpressure all sending subtasks on this multiplex:
+<br>
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2019-06-04-network-stack/flink-network-stack3.png" width="700px" alt="Physical-transport-backpressure-Flink's Network Stack"/>
```

Comment:
Is B.4 is causing the backpressure?

Reply:
I clarified this a bit better now / added a description of what the picture is showing